src/main/java/org/spongepowered/api/text/sink/CombinedSink.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.sink;
+
+import static org.apache.commons.lang3.Validate.noNullElements;
+
+import com.google.common.collect.ImmutableSet;
+import org.spongepowered.api.command.CommandSource;
+import org.spongepowered.api.text.Text;
+
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class CombinedSink implements MessageSink {
+
+    protected final Iterable<MessageSink> contents;
+
+    public CombinedSink(Iterable<MessageSink> contents) {
+        noNullElements(contents, "null message sink: %s");
+        this.contents = contents;
+    }
+
+    @Override
+    public Optional<Text> transformMessage(CommandSource recipient, Text original) {
+        Text text = original;
+        for (MessageSink sink : this.contents) {
+            text = sink.transformMessage(recipient, text).orElse(text);
+        }
+
+        return Optional.ofNullable(text);
+    }
+
+    @Override
+    public Iterable<CommandSource> getRecipients() {
+        return ImmutableSet.copyOf(StreamSupport.stream(this.contents.spliterator(), false)
+                .flatMap(sink -> StreamSupport.stream(sink.getRecipients().spliterator(), false))
+                .collect(Collectors.toSet()));
```

Comment:
IMO we should use a `ImmutableSet` builder collector here.

Just push something like this in a util class like `Functional`, so everyone can use it:

````java
public static <T> Collector<T, ImmutableSet.Builder<T>, ImmutableSet<T>> immutableSetCollector() {
    return Collector.of(ImmutableSet::builder, ImmutableSet.Builder::add,
            (left, right) -> left.addAll(right.build()), ImmutableSet.Builder::build,
            Collector.Characteristics.UNORDERED);
}
````

And then you can use it like this:
````java
return StreamSupport.stream(this.it.spliterator(), false)
        .flatMap(sink -> StreamSupport.stream(sink.getRecipients().spliterator(), false))
        .collect(immutableSetCollector());
````